hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java

```diff
@@ -134,6 +147,30 @@
         return plans;
     }
 
+    private static Address[] initPartitionOwnersAndMembers(NodeEngine nodeEngine, Collection<Member> members) {
+        Set<Address> addresses = members.stream().map(Member::getAddress).collect(toSet());
+        ClusterService clusterService = nodeEngine.getClusterService();
+
+        IPartitionService partitionService = nodeEngine.getPartitionService();
+        Address[] partitionOwners = new Address[partitionService.getPartitionCount()];
+        for (int partitionId = 0; partitionId < partitionService.getPartitionCount(); partitionId++) {
+            Address address = partitionService.getPartitionOwnerOrWait(partitionId);
+            partitionOwners[partitionId] = address;
+
+            if (!addresses.contains(address)) {
+                // Address in partition table doesn't exist in initial member list.
+                // Check whether it's a newly joined member?
+                Member member = clusterService.getMember(address);
+                if (member == null) {
+                    throw new TopologyChangedException("Topology changed! " + address + " is not member anymore!");
```

Comment:
If member leaves after this point, it's not different than member leaving while task is being executed. It will be handled by invocation system, invoking on a non-existent member will fail.

Logic here guarantees that, if an address is included in partition-owners list then it's going to exist in member-list too. But member list have members those are not in partition table, a member may not have any partition assignment at that moment or even that member may be a lite one.

> If the member wasn't in the initial member list but there is a partition assigned to it - then it must mean it's a new member.

It's either a new member or an old member that has just left but partition table is not updated yet.